pkg/registry/apps/deployment/strategy_test.go

```diff
@@ -41,6 +42,213 @@ const (
+			enableDeploymentPodReplacementPolicy: true,
+			disableDeploymentPodReplacementPolicyFGAfterCreate: true,
+			podReplacementPolicy:               ptr.To(apps.TerminationComplete),
+			expectedPodReplacementPolicy:       ptr.To(apps.TerminationComplete),
+			updatePodReplacementPolicy:         ptr.To(apps.TerminationStarted),
+			expectedUpdatePodReplacementPolicy: ptr.To(apps.TerminationStarted),
+		},
+		{
+			name:                                 "Deployment strategy with podReplacementPolicy should not be updated to invalid when DeploymentPodReplacementPolicy feature gate is disabled",
+			enableDeploymentPodReplacementPolicy: true,
+			disableDeploymentPodReplacementPolicyFGAfterCreate: true,
+			podReplacementPolicy:               ptr.To(apps.TerminationComplete),
+			expectedPodReplacementPolicy:       ptr.To(apps.TerminationComplete),
+			updatePodReplacementPolicy:         ptr.To(apps.DeploymentPodReplacementPolicy("Invalid")),
+			expectedValidationUpdateErr:        true,
+			expectedUpdatePodReplacementPolicy: ptr.To(apps.TerminationComplete),
+		},
+		{
+			name:                                 "Deployment strategy with podReplacementPolicy should be updated when both feature gates are disabled",
+			enableDeploymentPodReplacementPolicy: true,
+			disableDeploymentPodReplacementPolicyFGAfterCreate:          true,
+			disableDeploymentReplicaSetTerminatingReplicasFGAfterCreate: true,
+			podReplacementPolicy:               ptr.To(apps.TerminationComplete),
+			expectedPodReplacementPolicy:       ptr.To(apps.TerminationComplete),
+			updatePodReplacementPolicy:         ptr.To(apps.TerminationStarted),
+			expectedUpdatePodReplacementPolicy: ptr.To(apps.TerminationStarted),
+		},
+		{
+			name:                                 "Deployment strategy with podReplacementPolicy should not be updated to invalid when both feature gates are disabled",
+			enableDeploymentPodReplacementPolicy: true,
+			disableDeploymentPodReplacementPolicyFGAfterCreate:          true,
+			disableDeploymentReplicaSetTerminatingReplicasFGAfterCreate: true,
+			podReplacementPolicy:               ptr.To(apps.TerminationComplete),
+			expectedPodReplacementPolicy:       ptr.To(apps.TerminationComplete),
+			updatePodReplacementPolicy:         ptr.To(apps.DeploymentPodReplacementPolicy("Invalid")),
+			expectedValidationUpdateErr:        true,
+			expectedUpdatePodReplacementPolicy: ptr.To(apps.TerminationComplete),
+		},
+	}
+
+	withTwoFeatureGates := func(testName string, enableDeploymentPodReplacementPolicy bool, test func(testName string, deploymentReplicaSetTerminatingReplicasEnabled, deploymentPodReplacementPolicyEnabled bool)) {
+		for _, deploymentReplicaSetTerminatingReplicasEnabled := range []bool{true, false} {
+			for _, deploymentPodReplacementPolicyEnabled := range []bool{true, false} {
```

Comment:
I wonder, now that we have the ability to ensure FG dependencies, do we need to tests all variants of both FG on/off?